### PR TITLE
crates/sel4-capdl-initializer: add support for x86 IOPorts

### PR DESCRIPTION
**Pull request by Ivan-Velickovic**
Depends on https://github.com/coliasgroup/capdl/pull/2.

### CODE REVIEW

File: crates/sel4-capdl-initializer/core/src/lib.rs [hunk 1]
```diff
@@ -326,6 +326,23 @@ impl<'a, N: ObjectName, D: Content, M: GetEmbeddedFrame, B: BorrowMut<[PerObject
...
+                for (obj_id, start_port, end_port) in ioports {
+                    let slot = self.cslot_alloc_or_panic();
+                    init_thread::slot::IO_PORT_CONTROL
+                        .cap()
+                        .ioport_control_issue(start_port, end_port, &cslot_to_relative_cptr(slot))?;
+                    self.set_orig_cslot(obj_id, slot);
+                }
+            }
+        }
+
```
Comment (nspin) [inline crates/sel4-capdl-initializer/core/src/lib.rs:345]: Could you try this sort of conditional block instead?

```
#[sel4_cfg(ARCH_X86_64)]
{
  ...
}
```

If the compiler doesn't accept that, then could you do this, to wrap the whole thing in a new scope?

```
sel4::sel4_cfg_if! {
    if #[sel4_cfg(ARCH_X86_64)] {
        {
            let ...
        }
    }
}
```

Comment (nspin) [reply crates/sel4-capdl-initializer/core/src/lib.rs:345]: Also, could you move this block below the IrqHandler block beneath it?

**Review by nspin — CHANGES_REQUESTED** [review]

**Comment by Ivan-Velickovic** [comment]
Will `init_thread.rs` not fail if you're not compiling for an x86 system?

**Comment by Ivan-Velickovic** [comment]
It fails for me locally now yea.

**Comment by Ivan-Velickovic** [comment]
Maybe there needs to be something like:
```rust
        sel4_cfg_if! {
            #[cfg(ARCH_X86_64)]
            (IO_PORT_CONTROL, IOPortControl, seL4_CapIOPortControl),
        } else {
            (IO_PORT_CONTROL, Null, seL4_CapIOPortControl),
        }
```

Don't think I've got the syntax right but I think you'll know what I mean

File: crates/sel4/src/init_thread.rs [hunk 2]
```diff
@@ -134,8 +134,10 @@ pub mod slot {
         (IRQ_CONTROL, IrqControl, seL4_CapIRQControl),
         (ASID_CONTROL, AsidControl, seL4_CapASIDControl),
         (ASID_POOL, AsidPool, seL4_CapInitThreadASIDPool),
-        #[cfg(any())] // TODO
+        #[sel4_cfg(not(ARCH_X86_64))]
         (IO_PORT_CONTROL, Null, seL4_CapIOPortControl),
+        #[sel4_cfg(ARCH_X86_64)]
+        (IO_PORT_CONTROL, IOPortControl, seL4_CapIOPortControl),
```
Comment (Ivan-Velickovic) [inline crates/sel4/src/init_thread.rs:140]: This is how I've dealt with it, let me know if there's a better way.

**Comment by nspin** [comment]
Looks great, thanks!

**Pull request merged**